### PR TITLE
json output gets truncated when piped into another application

### PR DESCRIPTION
**Pull request by MerlinDMC**
This changes the `printHumanReadableReport` and `printJsonReport` functions to return promises so we can wait for their completion before we call `process.exit(1);` in case there is an audit problem.

When piping the JSON output into another unix application it will get truncated at 64k as `console.log` is async and does not wait for the whole buffer to be flushed.

Example:
```
node_modules/.bin/check-audit --json | wc -c
65536
```

The actual report is `371717` bytes in size:
```
node_modules/.bin/check-audit --json > audit-output.json ; ls -l audit-output.json
-rw-r--r-- 1 root root 371717 Jul 29 13:55 audit-output.json
```

Additionally I had to change `printJsonReport` to use a writeable stream instead of just `console.log` as it provides a callback as soon as the buffer gets flushed.

### CODE REVIEW

**Comment by MerlinDMC** [comment]
@naugtur Can we get this merged or should we look for other possible solutions to fix our audit pipeline failures outside of your package?

**Comment by naugtur** [comment]
Sorry for the inconvenience, I'll refund what you paid me for this software 😉

On a more serious note - it's not just a matter of clicking a button.

I've just finished a major rewrite and your PR is against the v1. I was looking into getting this introduced into both v1 and v2 while trying to maintain a full time job and a family. 

I'm likely to find time for this on Friday. Hope you'll be okay meanwhile.

**Comment by naugtur** [comment]
Making one of the functions async is a bit of a problem in v2. I think I know how to work around it. Could you share steps to reproduce? A package file that produces long output would be nice

**Comment by MerlinDMC** [comment]
@naugtur point me at the rewrite and I'll see if that works and if not will gladly provide a PR for that as well to make it a matter of clicking a button 😉 

I can't share a specific test case currently as those are internal ... but I might be able to arrange a package.json and package-lock.json pair that'll produce a large enough output.

The original problem tho is that `console.log` by default is NOT synchronous - which was the whole point of using the promises to chain functions...

**Comment by MerlinDMC** [comment]
https://gist.github.com/MerlinDMC/8fa0c8b3e5f6813f6d37962357ed2bef

That pair of `package.json` and `package-lock.json` produces a JSON blob of 68698 bytes so it should be cut of when piped into `wc -c`.

Steps to reproduce:
- pull the two files into a folder
- install via `npm --registry=https://registry.npmjs.org ci`
- `node_modules/.bin/check-audit --json > data.json ; ls -l data.json`
- `node_modules/.bin/check-audit --json | wc -c`

Results on my machine:

```
$ node_modules/.bin/check-audit --json > data.json ; ls -l data.json
-rw-r--r--  1 daniel  staff  68698 31 Jul 17:48 data.json
$ node_modules/.bin/check-audit --json | wc -c
65536
```

**Comment by danielabel** [comment]
Thanks @naugtur for looking at the PR. V happy to contribute time/effort on the codebase. How can I best help.

I’ve been wanting to blog about how we are building a view of our state of production and we are working to enable teams know and  to manage their risk... hmmm I guess i make it sound more than it is maybe, but I’d be happy to chat to you about it, before it gets blogged, if that is of use.

**Comment by naugtur** [comment]
Both of you please take a look at the 'refactoring' branch. It's missing one last step - I need to complete extracting core to a separate package.

I'm aware of the issues with console.log and honestly I didn't expect the output size. 
The refactoring branch has all output centralized and it was exactly for that reason, I anticipated console.log would fail at some point and require changing.
The views can be rewritten to async functions wrapping stdout.write - all if them for consistency.

I'm hoping to finish all work including this by the end of the week and release v2. Next step is to get npm and yarn to use the core module for respecting ignores.

Oh, and I'll gladly discuss use cases. I have some opinions and I'm using this project at work to manage my teams' products, all 30something of them.
Mail me to start conversation on how you use resolver.

File: check.js [hunk 1]
```diff
@@ -8,29 +8,35 @@ function auditOk(issues) {
 }
 
 function printHumanReadableReport(issues) {
-    if (auditOk(issues)) {
-        console.log("audit ok.");
-        return;
-    }
-    issues.forEach(issue => {
+    return new Promise((resolve) => {
```
Comment (naugtur) [inline check.js]: The more I look at the code, the more I think this fixing the issue is accidental. The promise is not really waiting for anything, buffers are flushed behind the scenes while Node.js runs another event loop cycle and if the output was long enough, it'd still end the process before all content is flushed.

File: check.js [hunk 2]
```diff
@@ -8,29 +8,35 @@ function auditOk(issues) {
...
-        `--------------------------------------------------`
-    );
+        resolve(issues);
+    });
 }
 
 function printJsonReport(issues) {
-    console.log(JSON.stringify(issues));
+    return new Promise((resolve) => {
+        process.stdout.write(JSON.stringify(issues), () => resolve(issues));
```
Comment (naugtur) [inline check.js]: This one is what I'd like to use everywhere. might wrap it a bit different

File: check.js (hunk 1 above)
Comment (naugtur) [reply check.js]: Or you did it just to make the function return a promise which was not necessary.

Comment (MerlinDMC) [reply check.js]: I made it return a promise to chain it in the then chain line 52.
It does not need to be waited for as the output seems to be in small enough chunks but it felt worse to have two print functions and only one is async.

File: check.js (hunk 2 above)
Comment (naugtur) [reply check.js]: oh well, I tried it in v2 and would have to rewrite and destabilize too much to make all printing asynchronous.

Comment (MerlinDMC) [reply check.js]:
```
const data = '{TEST TEST}'.repeat(1024 * 8);
if (process.stdout._handle) {
	process.stdout._handle.setBlocking(true);
}
console.log(data);
process.exit(0);
```

will also force it to wait for a buffer flush - but I'm not sure if that gives any other bad side effects.

**Comment by naugtur** [comment]
Check this out: https://github.com/naugtur/npm-audit-resolver/commit/82e6d18c079f3c9b7eba9efd2229f6d12ce2212d#diff-48aec7eaed476776d1d230f97c9e821bR2

A bit more overall testing (help welcome) and v2 is good to go

File: check.js (hunk 2 above)
Comment (naugtur) [reply check.js]: I went with async solution (in refactoring) for now, but making everything consistent and synchronous is also tempting.

**Comment by MerlinDMC** [comment]
Looks like that will fix it fine as well.
Do you want to push a 2.0 prerelease or should we build an internal package and run it across our apps to test it?
I will not take much effort to run them side by side and compare the output against the 1.x version.

**Comment by naugtur** [comment]
just published `npm-audit-resolver@next`

**Comment by MerlinDMC** [comment]
I'll prepare an A/B test and run it either over the weekend or early next week. Thx for your effort.

I'll leave this PR open till then to report back.

**Comment by MerlinDMC** [comment]
I ran the two versions against roughly 450 applications and checked for regressions - that all seems good.

I've checked if the truncation issue is fixed and it is as well in 2.0.

However it breaks the JSON output unfortunately.
Source of the break is https://github.com/naugtur/audit-resolver-core/blob/5bc96327a8351ac7ac8ddcd1c7ad159fc6aba6ee/views/main.js#L5

I think that should be written to stderr instead and we should be good.

**Comment by naugtur** [comment]
Thanks! Great work! 
I'll fix it tonight and republish both

**Comment by naugtur** [comment]
published v2.0.0-2 as `next`

**Comment by MerlinDMC** [comment]
```
root@04d6b3e4d32d:~/app# node_modules/.bin/check-audit --json | tee check.json | jq .metadata
You are using a deprecated file name. audit-resolv.json was renamed to audit-resolve.json. Any changes will be written to the new file and format.
{
  "totalDependencies": 12482,
  "optionalDependencies": 109,
  "devDependencies": 0,
  "dependencies": 12373,
  "vulnerabilities": {
    "critical": 0,
    "high": 5,
    "moderate": 0,
    "low": 0,
    "info": 0
  }
}
```

Works fine and doesn't truncate. I'll close this issue and we'll wait for 2.0 😃

**Comment by naugtur** [comment]
you can install `npm-audit-resolver@next` and keep it that way. I'm only going to add yarn support and release. Not expecting more changes unless someone finds bugs.

**Comment by MerlinDMC** [comment]
> Where did you get 450 apps? :)

We run some micro services ...

**Comment by danielabel** [comment]
> > Where did you get 450 apps? :)
> 
> We run some micro services ...

Yeah, I need to tell you what we are experimenting with, using the library this week :)
Last week was a crapshoot, apologies.

**Pull request closed without merging**